mesonbuild/environment.py

```diff
@@ -813,6 +813,12 @@ def detect_d_compiler(self, want_cross):
         # up to date language version at time (2016).
         if 'DC' in os.environ:
             exelist = shlex.split(os.environ['DC'])
+            for dc in exelist[:]:
```

Comment:
Wait, is this really a list of different executables not as other similar environment variables a list with an executable and additional arguments? That seems strange.

Reply:
`DC` is meant to be either an executable or a path to an executable + arguments to it. The example given by the poster is incorrect, but the implementation happens to be correct. For instance, you might have `DC='ccache gdc'`.

Reply:
Hum, now that it's morning and I'm looking at the patch again, `exelist.remove()` seems wrong, it should just raise an `EnvironmentException` and tell the user to fix the variable,

Reply:
If this in only one executable, wouldn't it be better to look at it's `--version` output below (i.e. line 831)?

Executables could have any name. Not sure if the version output actually contains enough information though.

Reply:
Yes, if we can do that, we should do that instead. This PR was merged too quickly.

Reply:
While I think aborting when an invalid D compiler is in there (as `DC` is indeed not supposed to be a "choose any of these compilers" list), I also think checking for executable names is justified here.
These wrappers can return all kinds of things for a version, and may even claim to be DMD directly, so detecting that will be rather annoying.
Checking for executable names is the easiest solution for 90% of all cases here.

Reply:
I apologise for that. When I wrote this patch, it was pretty late. I'm not very familiar with `Popen_*` API so I assumed `exelist` meant list of executables which was obviously wrong. I am surprised no one caught it when reviewing the PR. I'm sorry for the mess.

Reply:
Related to what @textshell suggested, it's impracticable, at least for `gdc`. `gdc --version` and `gdmd --version` return exactly the same output.

Reply:
That's a bit insane^^. Returning the exact same version and then doing something which would need special code when used.

But if we can not detect those wrappers by their version output, we are stuck with looking at the names.